enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,806 @@
+   Example 2: As a cluster administrator I want to set http request headers like `Content-Location "/my-first-blog-post"` , `X-SSL-Client-Cert "%{+Q}[ssl_c_der,base64]"` and `X-Target "%[req.hdr(host),lower]"` to be sent to all applications.
+   Example 3: As a cluster administrator I want to delete http response headers like `X-XSS-Protection` to be not sent to the client.
+   Example 4: As a cluster administrator I want to delete http request headers like `Content-Location` to be not sent to the application.
+2. As a user, I need to configure Route to set/delete HTTP request/response headers to send/not send them to the client which has called the route.
+   Example 1: As a user I want to set http response headers like `X-XSS-Protection "1; mode=block"` , `X-Content-Type-Options "nosniff"` and `X-Frame-Options "SAMEORIGIN"` to be sent to the client.
+   Example 2: As a user I want to set http request headers like `Content-Location "/my-first-blog-post"` , `X-SSL-Client-Cert "%{+Q}[ssl_c_der,base64]"` and `X-Target "%[req.hdr(host),lower]" to be sent to the application.`
+   Example 3: As a user I want to delete http response headers like `X-XSS-Protection` to be not sent to the client.
+   Example 4: As a user I want to delete http request headers like `Content-Location to be not sent to the application.`
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP response headers in the IngressController CR spec that should be set or
+   deleted at the global level when forwarding a response to all clients, i.e. for all routes.
+2. Enable the cluster administrator to specify HTTP request headers in the IngressController CR spec that should be set or
+   deleted at the global level when forwarding connections to all applications from clients.
+3. Enable the route owner to specify HTTP response headers in the Route spec (when it is not a passthrough Route) for their application
+   that should be set or deleted when forwarding a response to the client.
+4. Enable the route owner to specify HTTP request headers in the Route spec (when it is not a passthrough Route) for their application
+   that should be set or deleted when forwarding a connection to the application from a client.
+5. Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.
+   Eg- we should prohibit specifying strict-transport-security using the new API, for the following reasons:
+   HSTS is already configurable through a route annotation.
+   We already have an API for restricting HSTS in the cluster config.
+   Adding a second way to configure HSTS would significantly complicate this enhancement.
+6. Setting dynamic header values for http response. Allowed sample fetchers are res.hdr, ssl_c_der. Allowed converters are lower, base64.
+7. Setting dynamic header values for http request. Allowed sample fetchers are req.hdr, ssl_c_der. Allowed converters are lower, base64.
```

Comment:
Goals 6 and 7 were added after the RFE was accepted and tend to complicate this enhancement, as well as greatly increase time spent on code review, documentation, and QE.  Can we leave this as an iterative update for another time?

Reply:
This has been already considered now as some customer wanted it.